File: Algorithm/Python/695. Max Area of Island.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2017-10-08 21:59:05
# @Last Modified by:   WuLC
# @Last Modified time: 2017-10-08 21:59:19

# dfs
class Solution(object):
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if len(grid) == 0:
            return 0
        result = 0
        m, n = len(grid), len(grid[0])
        for i in xrange(m):
            for j in xrange(n):
                if grid[i][j] == 1:
                    result = max(result, self.dfs(grid, i, j))
        return result
    
    
    def dfs(self, grid, i, j):
        count = 1
        grid[i][j] = 0
        for di in [-1, 1]:   
            if 0 <= i + di < len(grid) and grid[i+di][j] == 1:
                count += self.dfs(grid, i + di, j)
        for dj in [-1, 1]:   
            if 0 <= j + dj < len(grid[0]) and grid[i][j+dj] == 1:
                count += self.dfs(grid, i , j + dj)
        return count
        
        
        

File: Algorithm/C++/695. Max Area of Island.cpp
/*
 * Created on Sat May 12 2018 19:50:30
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

// simple dfs
class Solution 
{
    public:
        int maxAreaOfIsland(vector<vector<int>>& grid) 
        {
            if(grid.size() == 0) return 0;
            int result = 0;
            int m = grid.size(), n = grid[0].size();
            for(int i = 0; i < m; i++)
            {
                for(int j = 0; j < n; j++)
                {
                    if (grid[i][j] == 1)
                        result = std::max(result, dfs(grid, i, j, m, n));
                }
            }
            return result;
        }
        
        int dfs(vector<vector<int>>& grid, int i, int j, int m, int n)
        {
            if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != 1) return 0;
            grid[i][j] = -1;
            return 1 + dfs(grid, i-1, j, m, n) + dfs(grid, i+1, j, m, n) + dfs(grid, i, j-1, m, n) + dfs(grid, i, j+1, m, n);
        }
};